cupyx/scipy/linalg/decomp_lu.py

```diff
@@ -59,7 +59,55 @@ def lu_factor(a, overwrite_a=False, check_finite=True):
         (array([[ 0.,  1.],
                [nan, nan]], dtype=float32), array([0, 1], dtype=int32))
     """
+    return _lu_factor(a, overwrite_a, check_finite)
 
+
+def lu(a, permute_l=False, overwrite_a=False, check_finite=True):
+    """LU decomposition.
+
+    Args:
+        a (cupy.ndarray): The input matrix with dimension ``(M, N)``
+        permute_l (bool): If ``True``, perform the multiplication ``P * L``.
+        overwrite_a (bool): Allow overwriting data in ``a`` (may enhance
+            performance)
+        check_finite (bool): Whether to check that the input matrices contain
+            only finite numbers. Disabling may give a performance gain, but may
+            result in problems (crashes, non-termination) if the inputs do
+            contain infinities or NaNs.
+        a (cupy.ndarray): Array to decompose
+
+    Returns:
+        **(If permute_l == False)**
+        p (cupy.ndarray): Permutation matrix with dimension ``(M, M)``
+        l (cupy.ndarray): Lower triangular or trapezoidal matrix with unit
+            diagonal with dimension ``(M, K)``. ``K = min(M, N)``.
+        u (cupy.ndarray): Upper triangular or trapezoidal matrix with dimension
+            ``(K, N)``. ``K = min(M, N)``.
+
+        **(If permute_l == True)**
+        pl (cupy.ndarray): Permuted ``L`` matrix. ``K = min(M, N)``.
+        u (cupy.ndarray): Upper triangular or trapezoidal matrix with dimension
+            ``(K, N)``. ``K = min(M, N)``.
+
+    .. seealso:: :func:`scipy.linalg.lu`
+    """
+    lu, piv = _lu_factor(a, overwrite_a, check_finite)
+
+    m, n = lu.shape
+    k = min(m, n)
+    L, U = cupy_split_lu(lu)
+
+    if permute_l:
+        cupy_laswp(L, 0, k-1, piv, -1)
+        return (L, U)
+    else:
+        r_dtype = numpy.float32 if lu.dtype.char in 'fF' else numpy.float64
+        P = cupy.diag(cupy.ones((m,), dtype=r_dtype))
+        cupy_laswp(P, 0, k-1, piv, -1)
+        return (P, L, U)
+
+
+def _lu_factor(a, overwrite_a=False, check_finite=True):
```

Comment:
In #4001, let's discuss the public/private LU factorization interface.  OK to leave the issue in this PR.